runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java

```diff
@@ -523,12 +563,22 @@ public void invalidateClientCache() {
      * @param addresses     collection of addresses to read from.
      * @param waitForWrite  flag whether wait for write is required or hole fill directly.
      * @param cacheOnServer flag whether read results should be cached on log unit server.
+     * @param cacheOnClient flag whether read will be cached in the client.
+     * @param ignoreTrimmed flag whether trimmed exceptions are ignored or not.
+     *
+     *                      If trimmed exceptions are ignored the fetchAll will
+     *                      return the non-trimmed addresses from the requested set,
+     *                      otherwise, in the event of trimmed exceptions, non-trimmed addresses
+     *                      will be cached and a trim exception with the specified trimmed addresses
+     *                      will be thrown.
      * @return an ordered map of read addresses.
      */
     @Nonnull
-    public Map<Long, ILogData> fetchAll(Iterable<Long> addresses,
-                                        boolean waitForWrite,
-                                        boolean cacheOnServer) {
+    private Map<Long, ILogData> fetchAll(Iterable<Long> addresses,
+                                         boolean waitForWrite,
+                                         boolean cacheOnServer,
+                                         boolean cacheOnClient,
```

Comment:
fetchAll() by definition should not deal with caching on client, it's the caller of it, i.e. read(), doing this, we shouldn't mix them up, right?

Reply:
I pushed it down to the fetch all, as the checkLogData (check for trimmed addresses) is the one throwing the TrimmedException and for multiple reads it is convenient to cache the valid reads before throwing back the TrimmedException (in the case the flag is not set),  or we will be going again to the log units for these set of valid addresses. I wanted to avoid requiring fetchAll + checkLogData (2 calls) and confine them through the fetchAll only (as it is also used by the nonCacheFetchAll) 

If you consider the fetchAll should not deal with caching, I can take the checkLogData out of this call and call it after the fetchAll... I. was avoiding the combination of these two...

Reply:
I think taking the checkLogData out of fetchAll() might not be good because there are someone else consuming fetchAll() directly like state transfer/fast loader, then you have to add the check to consumer's logic, which isn't good. Maybe we can populate TrimmedException up to the read() API and let read() catch and handle it as well?

Reply:
That is precisely why I left it within the fetchAll... the problem with bubbling it up to the read api is that if you throw a trimmedException you lose the valid LogData entries, unless we push them in the TrimmedException but I thought this is not really clean. Any thoughts?

Reply:
In checkForTrimmedAddresses(), you will put trimmedAddress into TrimmedException, so in read() when you get the exception, you know the trimmedAddress, so then you can cache the addresses that are not trimmed, right?